### PR TITLE
Use intra-doc links in core/src/iter when possible

### PR DESCRIPTION
**Pull request by poliorcetics**
Helps with #75080.

I also updated lots of links to use `fn()` instead of `fn` when possible.

@rustbot modify labels: T-doc A-intra-doc-links

r? @jyn514

### CODE REVIEW

File: library/core/src/iter/sources.rs [hunk 1]
```diff
@@ -601,10 +581,8 @@ where
 
 /// An new iterator where each successive item is computed based on the preceding one.
 ///
-/// This `struct` is created by the [`successors`] function.
+/// This `struct` is created by the [`successors()`] function.
```
Comment (jyn514) [inline library/core/src/iter/sources.rs:584]:
```suggestion
/// This `struct` is created by the [`iter::successors()`] function.
```

**Comment by pickfire** [comment]
@poliorcetics I have a patch for sources too https://github.com/rust-lang/rust/pull/76816 but I was inclined to move them to `fn()` since I just put them as `fn`.

**Comment by jyn514** [comment]
@bors r+ rollup

It would be nice to move from 'the function f' to 'f()' at some point, but that doesn't need to block this PR.

File: library/core/src/iter/sources.rs [hunk 2]
```diff
@@ -393,10 +379,8 @@ pub fn once<T>(value: T) -> Once<T> {
 /// An iterator that yields a single element of type `A` by
 /// applying the provided closure `F: FnOnce() -> A`.
 ///
-/// This `struct` is created by the [`once_with`] function.
+/// This `struct` is created by the [`once_with()`] function.
```
Comment (pickfire) [inline library/core/src/iter/sources.rs:382]:
```suggestion
/// This `struct` is created by [`once_with()`].
```

**Pull request merged**